diff --git a/dist/runtime-vapor.esm-bundler.js b/dist/runtime-vapor.esm-bundler.js
--- a/dist/runtime-vapor.esm-bundler.js
+++ b/dist/runtime-vapor.esm-bundler.js
@@ -2940,6 +2940,10 @@
 		devRender(instance);
 	}
 	else if (setupResult === EMPTY_OBJ && component.render) instance.block = callRender(component.render, instance, setupResult);
+	else if (!isBlock(setupResult) && component.render) {
+		instance.setupState = proxyRefs(setupResult);
+		instance.block = callRender(component.render, instance, instance.setupState);
+	}
 	else instance.block = setupResult;
 	if (instance.hasFallthrough && component.inheritAttrs !== false && Object.keys(instance.attrs).length) {
 		const getFallthroughAttrs = isFunction(component) && !(isTransitionEnabled ? isVaporTransition(component) : false) ? () => getFunctionalFallthrough(instance.attrs) : () => instance.attrs;
@@ -6384,7 +6388,7 @@
 */
 function setRef$1(instance, el, ref, oldRef, refFor = false, refKey, oldRefKey) {
 	if (!instance || instance.isUnmounted) return;
-	const setupState = !!(process.env.NODE_ENV !== "production") ? instance.setupState || {} : null;
+	const setupState = instance.setupState || {};
 	const refValue = getRefValue(el);
 	if (isInteropEnabled) {
 		const target = isFragment(el) && el.setRef ? el : refValue && isFragment(refValue) && refValue.setRef ? refValue : null;
@@ -6394,7 +6398,7 @@
 		}
 	}
 	const refs = instance.refs === EMPTY_OBJ ? instance.refs = {} : instance.refs;
-	const canSetSetupRef = !!(process.env.NODE_ENV !== "production") ? createCanSetSetupRefChecker(setupState, refs) : NO;
+	const canSetSetupRef = createCanSetSetupRefChecker(setupState, refs);
 	const canSetRef = (ref, key) => {
 		if (!!(process.env.NODE_ENV !== "production") && knownTemplateRefs.has(ref)) return false;
 		if (key && isTemplateRefKey(refs, key)) return false;
@@ -6404,7 +6408,7 @@
 		invalidatePendingRef(el);
 		if (isString(oldRef)) {
 			refs[oldRef] = null;
-			if (!!(process.env.NODE_ENV !== "production") && canSetSetupRef(oldRef)) setupState[oldRef] = null;
+			if (canSetSetupRef(oldRef)) setupState[oldRef] = null;
 		} else if (isRef$1(oldRef)) {
 			if (canSetRef(oldRef, oldRefKey)) oldRef.value = null;
 			if (oldRefKey) refs[oldRefKey] = null;
@@ -6428,12 +6432,12 @@
 			const doSet = () => {
 				if (refFor) {
 					if (refValue == null) return;
-					existing = _isString ? !!(process.env.NODE_ENV !== "production") && canSetSetupRef(ref) ? setupState[ref] : refs[ref] : canSetRef(ref) || !refKey ? ref.value : refs[refKey];
+					existing = _isString ? canSetSetupRef(ref) ? setupState[ref] : refs[ref] : canSetRef(ref) || !refKey ? ref.value : refs[refKey];
 					if (!isArray(existing)) {
 						existing = [refValue];
 						if (_isString) {
 							refs[ref] = existing;
-							if (!!(process.env.NODE_ENV !== "production") && canSetSetupRef(ref)) {
+							if (canSetSetupRef(ref)) {
 								setupState[ref] = refs[ref];
 								existing = setupState[ref];
 							}
@@ -6444,7 +6448,7 @@
 					} else if (!existing.includes(refValue)) existing.push(refValue);
 				} else if (_isString) {
 					refs[ref] = refValue;
-					if (!!(process.env.NODE_ENV !== "production") && canSetSetupRef(ref)) setupState[ref] = refValue;
+					if (canSetSetupRef(ref)) setupState[ref] = refValue;
 				} else if (_isRef) {
 					if (canSetRef(ref, refKey)) ref.value = refValue;
 					if (refKey) refs[refKey] = refValue;
@@ -6456,7 +6460,7 @@
 					if (isArray(existing)) remove$1(existing, refValue);
 				} else if (_isString) {
 					refs[ref] = null;
-					if (!!(process.env.NODE_ENV !== "production") && canSetSetupRef(ref)) setupState[ref] = null;
+					if (canSetSetupRef(ref)) setupState[ref] = null;
 				} else if (_isRef) {
 					if (canSetRef(ref, refKey)) ref.value = null;
 					if (refKey) refs[refKey] = null;
